lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3653,17 +3654,19 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     return (insertion.first->getSecond() = bufWithCleanup);
   }
 
-  void addToAdjointBuffer(SILValue originalBuffer,
+  void addToAdjointBuffer(SILValue destBuffer,
                           SILValue newValueBufferAccess) {
-    assert(originalBuffer->getType().isAddress() &&
+    assert(destBuffer->getType().isAddress() &&
            newValueBufferAccess->getType().isAddress());
-    auto buf = getAdjointBuffer(originalBuffer);
-    auto *access = builder.createBeginAccess(
-        newValueBufferAccess.getLoc(), buf, SILAccessKind::Modify,
+    assert(destBuffer->getFunction() == &getAdjoint());
+    assert(newValueBufferAccess->getFunction() == &getAdjoint());
+    auto *destAccess = builder.createBeginAccess(
+        newValueBufferAccess.getLoc(), destBuffer, SILAccessKind::Modify,
```

Comment:
Note: I think it's now wrong to always initialize `destAccess` with ` SILAccessKind::Modify`. `SILAccessKind::Init` should be used if `destBuffer` hasn't been initialized before.

(Previously, `Modify` was correct because `begin_access` was created on the output of `getAdjointBuffer(originalBuffer)`, which initializes adjoint buffers to zero and calls  `begin_access [init]`.)

I believe we can detect whether to use `Init` or `Modify` by checking if there exists a use of `destBuffer` that is a `begin_access [init]`.

@rxwei: what are your thoughts?

Reply:
I don't understand why you changed `addToAdjointBuffer`'s argument's role. I'd expect it to have the old behavior.

Reply:
Resolved in [the other comment thread](https://github.com/apple/swift/pull/22669#discussion_r257488408).